spring-kafka/src/main/java/org/springframework/kafka/support/SeekUtils.java

```diff
@@ -0,0 +1,108 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.support;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiPredicate;
+
+import org.apache.commons.logging.Log;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.TopicPartition;
+
+/**
+ * Seek utilities.
+ *
+ * @author Gary Russell
+ * @since 2.2
+ *
+ */
+public final class SeekUtils {
+
+	/**
+	 * The number of times a topic/partition/offset can fail before being rejected.
+	 */
+	public static final int DEFAULT_MAX_FAILURES = 10;
+
+	private SeekUtils() {
+		super();
+	}
+
+	/**
+	 * Seek records to earliest position, optionally skipping the first.
+	 * @param records the records.
+	 * @param consumer the consumer.
+	 * @param exception the exception
+	 * @param recoverable true if skipping the first record is allowed.
+	 * @param skipper function to determine whether or not to skip seeking the first.
+	 * @param logger a {@link Log} for seek errors.
+	 * @param <K> the key type.
+	 * @param <V> the value type.
+	 */
+	public static  <K, V> void doTypedSeeks(List<ConsumerRecord<K, V>> records, Consumer<K, V> consumer, Exception exception,
+			boolean recoverable, BiPredicate<ConsumerRecord<K, V>, Exception> skipper, Log logger) {
+		Map<TopicPartition, Long> partitions = new LinkedHashMap<>();
+		AtomicBoolean first = new AtomicBoolean(true);
+		records.forEach(record ->  {
+			if (!recoverable || !first.get() || !skipper.test(record, exception)) {
+				partitions.computeIfAbsent(new TopicPartition(record.topic(), record.partition()), offset -> record.offset());
+			}
+			first.set(false);
+		});
+		partitions.forEach((topicPartition, offset) -> {
+			try {
+				consumer.seek(topicPartition, offset);
+			}
+			catch (Exception e) {
+				logger.error("Failed to seek " + topicPartition + " to " + offset, e);
+			}
+		});
+	}
+
+	/**
+	 * Seek records to earliest position, optionally skipping the first.
+	 * @param records the records.
+	 * @param consumer the consumer.
+	 * @param exception the exception
+	 * @param recoverable true if skipping the first record is allowed.
+	 * @param skipper function to determine whether or not to skip seeking the first.
+	 * @param logger a {@link Log} for seek errors.
+	 */
+	public static void doSeeks(List<ConsumerRecord<?, ?>> records, Consumer<?, ?> consumer, Exception exception,
```

Comment:
I don't really like having 2 methods here; the problem is one client has generic types, the other doesn't.

One solution would be to remove the generic types from `AfterRollbackPostProcessor` but I wonder if that's too much of a breaking change? Perhaps not.

WDYT?